Added gaussian_blur_op

ghosalsattam opened the pull request:
This PR is aimed at putting a gaussian blur in an image. It reduces the noise in an image.
See Issue #1333

ghosalsattam commented:
@seanpmorgan @gabrieldemarmiesse @WindQAQ @autoih Please provide a review

Squadrick commented:
@ghosalsattam This is PR requires a lot of work to get to an acceptable state:

1. You can't put the file in the base directory, put in under `tensorflow_addons/image`. 
2. Update `BUILD` in `tensorflow_addons/image`.
3. The naming scheme of files should be `file_name.py`, not `fileName.py`. 
4. You've implemented the op as an object, whereas it should actually be implemented as a function. See some of the other ops in `tensorflow_addons/image` to get a better idea. 
5. You should use `tf` for the implementation instead of `np`. (`ones`, `expand_dims`). 
6. All core TF ops should be used via the public API. So use `tf.nn.convolution` directly. Remove all `tensorflow.python.*` imports.
7. The testing code should run using `pytest`. 
8. The test should compare the output with other standard image processing libraries like PIL or sklearn-image.

ghosalsattam commented:
@Squadrick I am trying to use tf.Tensor to find Gaussian Kernel.But when I am using the function assign in a variable of type tf.Variable I am getting the error 'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'assign'
Can you please help me with the problem as I am new to tensorflow open source community. 
Thank You

Squadrick commented:
@ghosalsattam Can you post the code that's causing the error.

ghosalsattam commented:
Ok

ghosalsattam commented:
def findKernel(sigma,kSize,gaussianFilter):
		"This function creates a kernel of size [kSize*kSize]"
		for i in range(-kSize//2,kSize//2+1):
			for j in range(-kSize//2,kSize//2+1):
				gaussianFilter[i+kSize//2][j+kSize//2].assign(1/(2*np.pi*(sigma)**2)*np.exp(-(i**2+j**2)/(2*sigma**2)))
		print(gaussianFilter)
		return

gaussianFilter=tf.Variable(tf.zeros([5,5]))
findKernel(1,5,gaussianFilter)

ghosalsattam commented:
`def findKernel(sigma,kSize,gaussianFilter):
		"This function creates a kernel of size [kSize*kSize]"
		for i in range(-kSize//2,kSize//2+1):
			for j in range(-kSize//2,kSize//2+1):
				gaussianFilter[i+kSize//2][j+kSize//2].assign(1/(2*np.pi*(sigma)**2)*np.exp(-(i**2+j**2)/(2*sigma**2)))
		print(gaussianFilter)
		return

gaussianFilter=tf.Variable(tf.zeros([5,5]))

findKernel(1,5,gaussianFilter)

ghosalsattam commented:
@Squadrick Please help with the code I provided above

ghosalsattam commented:
Ok I solved it. Thank you for the support and sorry for the inconvenience caused.

ghosalsattam commented:
@Squadrick plaese provide a review.

ghosalsattam commented:
@seanpmorgan @gabrieldemarmiesse @WindQAQ @autoih @Squadrick  Please provide a review

ghosalsattam commented:
Fixes Issue #1353

gabrieldemarmiesse commented:
@ghosalsattam we usually provide review when all the tests in the CI are passing. Here you have some failed tests. You can look at the logs by clicking on "details".

ghosalsattam commented:
@Squadrick @gabrieldemarmiesse One test case is not running Build release wheels (ubuntu-18.04, 3.5) (pull_request) Failing after 9m — Build release wheels (ubuntu-18.04, 3.5)
Can you help me with the issue.

Squadrick commented:
@ghosalsattam The failure is due to the added dep of OpenCV. Since we already use `PIL` for other tests, why not use `PIL.ImageFilter.GaussianBlur` as the ground truth for testing your function?

ghosalsattam commented:
@ghosalsattam @gabrieldemarmiesse Please give a review

ghosalsattam commented:
@Squadrick @gabrieldemarmiesse Please provide a review

ghosalsattam commented:
@Squadrick @gabrieldemarmiesse @facaiy Please provide a review

gabrieldemarmiesse commented:
@ghosalsattam Maintainers are nearly all volunteers working on Addons during their free time, and the list of pull requests is quite long, it may take a few days, sometimes weeks to get a review depending on the size of the diff. No worries, as long as the build is passing and the pull request is opened, you'll get a review at some point.

ghosalsattam commented:
@Squadrick I messed up pushing commits n the master branch. Now I want to create new branch and push commits. But I am unable to do so because master branch has commits. When I am making new PR from the new branches the PR contain the older commits.I am having few more files fixing other issues. Can you help me out of the situation.

ghosalsattam commented:
@Squadrick Any More changes you suggest?

gabrieldemarmiesse commented on tools/install_deps/pytest.txt:
```diff
@@ -3,4 +3,5 @@ pytest-xdist~=1.31
 scikit-learn~=0.22
 scikit-image~=0.15.0
 Pillow~=7.0.0
+opencv-python~=4.1.1
```
It seems opencv is not used in your pull request, you can remove it.

ghosalsattam replied:
Removed!!!

gabrieldemarmiesse commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,54 @@
+    assert accuracy >= 0.80
+
+
+if __name__ == "__main__":
+    sys.exit(pytest.main([__file__]))
```
It's not needed in our new test setup.

ghosalsattam replied:
Done!!!

gabrieldemarmiesse commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,54 @@
+
+
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+def test():
```
Test functions should start with `test_`, you can take a look at our new test guide: https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#writing-tests

ghosalsattam replied:
Done

gabrieldemarmiesse commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,54 @@
+
+    accuracy = 0
+
+    for i in range(len(gb1)):
+        for j in range(len(gb1[0])):
+            if abs(gb1[i][j] - gb2[i][j]) < 10:
+                accuracy += 1
```
Please use the `np.testing` module instead of computing the difference by hand. The error messages are much better.

gabrieldemarmiesse commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,70 @@
+def findKernel(sigma, kSize, gaussianFilter):
+    "This function creates a kernel of size [kSize*kSize]"
+    rowFilter = tf.Variable(tf.zeros(kSize, dtype=tf.float64), dtype=tf.float64)
+    for i in range(-kSize // 2, kSize // 2 + 1):
+        for j in range(-kSize // 2, kSize // 2 + 1):
+            rowFilter[j + kSize // 2].assign(
+                1
+                / (2 * np.pi * (sigma) ** 2)
+                * tf.exp(tf.cast(-(i ** 2 + j ** 2) / (2 * sigma ** 2.00), tf.float64))
+            )
+        gaussianFilter[i + kSize // 2].assign(rowFilter)
```
This is going to be too slow. Please vectorize those operations rather than doing for loops.

ghosalsattam replied:
Addressed

gabrieldemarmiesse commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,70 @@
+    return conv_ops
+
+
+def findKernel(sigma, kSize, gaussianFilter):
```
Please respect pep8 conventions concerning naming.

ghosalsattam commented:
@gabrieldemarmiesse @Squadrick   I have implemented all the changes you suggsted.
But I am getting few errors which were not coming earlier. All of them related to publish wheels PyPi.
Can You help. Link-https://github.com/ghosalsattam/addons/

gabrieldemarmiesse commented:
@ghosalsattam Pushing to pypi requires a login and password that is only available in the original addons, not it's normal that it fails for this step in your fork. I'll re-open the pull request, that should work out.

ghosalsattam commented:
@gabrieldemarmiesse I have implemented a gaussian kernel which is sequence of 2 1d convolution one in x direction and other in y-direction. This is implemented in other libraries like scipy, opencv etc as this method is faster. All the tests are running as suggested by you. Please provide a review.

ghosalsattam commented:
@gabrieldemarmiesse any more changes you suggest.

ghosalsattam commented:
@gabrieldemarmiesse Please provide a review or any more suggestions.

WindQAQ commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,63 @@
+
+
+def gaussian_blur(
+    img: FloatTensorLike, sigma: tf.float64, kSize: tf.float64
```
`FloatTensorLike` for `sigma`, and `int` for `kernel_size`.

ghosalsattam replied:
Done!!

WindQAQ commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,63 @@
+    which follows normal distribution then convolves the image with the kernel.Presently it works only on
+    grayscale images.
+    Args:
+    img: A tensor of shape
```
`image` should be a better name.

ghosalsattam replied:
Done

WindQAQ commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,63 @@
+    sigma:A constant of type float64. It is the standard deviation of the normal distribution.
+          The more the sigma, the more the blurring effect.
+          G(x,y)=1/(2*3.14*sigma**2)e^((x**2+y**2)/2sigma**2)
+    kSize:It is the kernel-size for the Gaussian Kernel. kSize should be odd.
```
`kernel_size` should be a better name.

ghosalsattam replied:
Done

WindQAQ commented on tensorflow_addons/image/gaussian_filter_ops.py:
```diff
@@ -0,0 +1,63 @@
+        a = tf.reshape(a, [kSize, 1, 1, 1])
+    else:
+        a = tf.reshape(a, [1, kSize, 1, 1])
+    return a
```
I would recommend
```python
def gaussian_kernel(kernel_size, sigma):
    # compute 1d kernel
```
and put the reshaping and casting stuff to `gaussian_filter2d`.

ghosalsattam replied:
Done

WindQAQ commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,41 @@
+
+from tensorflow_addons.image.gaussian_filter_ops import gaussian_blur
+import numpy as np
+import numpy.core.multiarray
```
this is not used.

ghosalsattam replied:
Done

WindQAQ commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,41 @@
+    test_image_cv = np.resize(test_image_cv, [40, 40])
+    gb2 = gaussian_filter(test_image_cv, 1, truncate=4.6, mode="constant")
+    print(gb1)
+    print(gb2)
```
please remove unnecessary print.

ghosalsattam replied:
Done

WindQAQ commented on tensorflow_addons/image/tests/gaussian_filter_ops_test.py:
```diff
@@ -0,0 +1,41 @@
+    print(gb2)
+    gb1 = gb1[:, :] / 255
+    gb2 = gb2[:, :] / 255
+    np.testing.assert_allclose(gb2, gb1, 0.5)
```
Why we need to use such a large tolerance 0.5?

ghosalsattam replied:
Done

ghosalsattam commented:
@WindQAQ @gabrieldemarmiesse @Squadrick I have done all the changes suggested. Kindly provide a review at the earliest.

mihaimaruseac commented on .github/CODEOWNERS:
```diff
@@ -58,8 +58,8 @@
 /tensorflow_addons/image/tests/distort_image_ops_test.py @windqaq
-/tensorflow_addons/image/filters.py @mainak431
-/tensorflow_addons/image/tests/filters_test.py @mainak431
+/tensorflow_addons/image/filters.py @mainak431 @ghosalsattam
+/tensorflow_addons/image/tests/filters_test.py @mainak431 @ghosalsattam
```
These changes should be reverted. The `CODEOWNERS` file contains people from the team at Google that know everything about the code in the corresponding entries.

Contributors can contribute to code without being listed in `OWNERS` files

ghosalsattam replied:
@mihaimaruseac I have removed it.

seanpmorgan replied:
@mihaimaruseac For TensorFlow Addons, maintainers are community members (typically not Googlers). For new contributions we require contributors be willing to maintain the code and add themselves to the CODEOWNERS. Perhaps mistaken which repository this PR was for?

@ghosalsattam Please revert your revert  :laughing:

ghosalsattam replied:
@seanpmorgan I have added the codeowners again.

seanpmorgan commented on tensorflow_addons/image/__init__.py:
```diff
@@ -35,3 +35,4 @@
 from tensorflow_addons.image.transform_ops import transform
 from tensorflow_addons.image.translate_ops import translate
 from tensorflow_addons.image.translate_ops import translate_xy
+from tensorflow_addons.image.filters import gaussian_filter2d
```
Please put this in alphabetical order

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,57 @@ def median_filter2d(
+def gaussian_filter2d(
+    image: FloatTensorLike,
+    sigma: FloatTensorLike,
+    kernel_size: int,
```
Please utilize the same parameter `filter_shape` as the other functions in the `filters.py` module. We need to be consistent so users can switch filters and not have a poor user experience.

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,57 @@ def median_filter2d(
+    sigma:A constant of type float64. It is the standard deviation of the normal distribution.
+          The more the sigma, the more the blurring effect.
+          G(x,y)=1/(2*3.14*sigma**2)e^((x**2+y**2)/2sigma**2)
+    kSize:It is the kernel-size for the Gaussian Kernel. kSize should be odd.
```
This documentation needs to be updated: 
* kSize -> filter_shape
* Documentation on `padding` parameter
* Documentation on `constant_values` parameter

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,57 @@ def median_filter2d(
+        raise ValueError("Sigma should not be zero")
+    if padding not in ["REFLECT", "CONSTANT", "SYMMETRIC"]:
+        raise ValueError("Padding should be REFLECT, CONSTANT, OR SYMMETRIC")
+
```
Can you implement the `img_utils.to_4D_image` function that is used in the other filters. Unfortunately we can't merge a single channel only function because it's so likely that it would be used for more than just that.

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,57 @@ def median_filter2d(
+    image = _pad(
+        image, (kernel_size, kernel_size), mode=padding, constant_values=constant_values
+    )
+    conv_ops_x = tf.nn.convolution(
+        input=image, filters=gaussian_filter_x, padding="VALID"
+    )
+    conv_ops = tf.nn.convolution(
+        input=conv_ops_x, filters=gaussian_filter_y, padding="VALID"
+    )
```
You'll likely want to use depthwise_conv2d in order to support multi-channel images:
https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,57 @@ def median_filter2d(
+    return conv_ops
+
+
+def find_kernel(sigma, kernel_size):
+    "This function creates a kernel of size [kSize]"
+    x = tf.range(-kernel_size // 2 + 1, kernel_size // 2 + 1)
+    x = tf.math.square(x, tf.float64)
+    a = tf.exp(-(x) / (2 * (sigma ** 2)))
+    a = a / tf.math.reduce_sum(a)
+    return a
```
This method should be private (i.e. leading with `_`). Also it's probably better to just make this a `_get_gaussian_kernel` method. This gist may be a good starting point:
https://gist.github.com/blzq/c87d42f45a8c5a53f5b393e27b1f5319

ghosalsattam replied:
Done

seanpmorgan commented on tensorflow_addons/image/tests/filters_test.py:
```diff
@@ -359,3 +361,45 @@ def test_symmetric_padding_with_3x3_filter_median(image_shape):
         expected_plane=expected_plane,
     )
+
+
```
Please add test cases for 4D images. With support for 'NHWC' or 'NCHW' data formats

ghosalsattam replied:
Done

ghosalsattam commented:
@seanpmorgan I have done all the changes suggested by you. Please provide a review.
I just have one query in my mind. First I had opened the PR at tensorflow/tensorflow repository as GSOC contribution. There I was asked to open the PR here. But I have seen tf.image which containg similar operations and Gaussian_Blur() functionality is missing even over there. Is it that if the PRs opened here, if they are working properly are migrated to tensorflow main repository later. I am a new guy to tensorflow/addons and sorry if the question is not making sense. Thank you for the help.

AakashKumarNain commented on tensorflow_addons/image/tests/filters_test.py:
```diff
@@ -359,3 +362,111 @@ def test_symmetric_padding_with_3x3_filter_median(image_shape):
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+def test_gaussian_filter2d_constant():
+    test_image_tf = tf.random.uniform(
+        [1, 40, 40, 1], minval=0, maxval=255, dtype=tf.float64
```
Why `float64`? IMO `float32` would be fine. Because everything will be in `float32`, it will be casted to float32 anyways in that case

ghosalsattam replied:
Actually in the exponentiation, I am using float64. So to keep the data type uniform over all the operations I am using double.

AakashKumarNain replied:
Is `float32` causing overflow?

ghosalsattam replied:
Done

AakashKumarNain commented on tensorflow_addons/image/tests/filters_test.py:
```diff
@@ -359,3 +362,111 @@ def test_symmetric_padding_with_3x3_filter_median(image_shape):
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+def test_gaussian_filter2d_reflect():
+    test_image_tf = tf.random.uniform(
+        [1, 40, 40, 1], minval=0, maxval=255, dtype=tf.float64
```
Same ques regarding float64

ghosalsattam replied:
Can you please elaborate?

AakashKumarNain commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,78 @@ def median_filter2d(
+def _get_gaussian_kernel(sigma, filter_shape, channels):
+    "This function creates a kernel of size [filter_shape]"
+    x = tf.range(-filter_shape // 2 + 1, filter_shape // 2 + 1)
+    x = tf.math.square(x, tf.float64)
```
Until unless there is an overflow when using`float32`, we should stick to `float32`. If there is any other reason, please let us know

ghosalsattam replied:
Addressed!!!

AakashKumarNain commented:
@ghosalsattam just few more changes in the main file. Almost there!

ghosalsattam commented:
> @ghosalsattam just few more changes in the main file. Almost there!

@AakashKumarNain what more changes are you suggesting?

AakashKumarNain commented:
> > @ghosalsattam just few more changes in the main file. Almost there!
> 
> @AakashKumarNain what more changes are you suggesting?

Check my comments in ` tensorflow_addons/image/filters.py ` file. I will put the question here as well. 
```python
def _get_gaussian_kernel(sigma, filter_shape, channels):
    "This function creates a kernel of size [filter_shape]"
    x = tf.range(-filter_shape // 2 + 1, filter_shape // 2 + 1)
    x = tf.math.square(x, tf.float64)
```

Is there any reason why we are making `float64` default here? Is there any overflow? If not, I would suggest to change this to `float32` as well

seanpmorgan commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,87 @@ def median_filter2d(
+def gaussian_filter2d(
+    image: FloatTensorLike,
+    sigma: FloatTensorLike,
+    filter_shape: int,
```
Please match the input of the other filters. Integer or tuple so that users can specify non-square kernels

ghosalsattam replied:
Addressed!

AakashKumarNain commented:
Thanks LGTM! @seanpmorgan thoughts?

autoih commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+        (batch_size, height, width, channels)
+        (NHWC), (batch_size, channels, height, width)(NCHW).
+
+    sigma:A constant of type float64. It is the standard deviation of the normal distribution.
```
sigma:A constant of type float64 -> float32 ?

ghosalsattam commented:
@seanpmorgan @autoih @AakashKumarNain I have updated the files addressing the issues raised by you. Is it in a state to be merged or need some more refinements.

AakashKumarNain commented on tensorflow_addons/image/__init__.py:
```diff
@@ -21,6 +21,7 @@
 from tensorflow_addons.image.cutout_ops import cutout
 from tensorflow_addons.image.dense_image_warp import dense_image_warp
 from tensorflow_addons.image.distance_transform import euclidean_dist_transform
+from tensorflow_addons.image.filters import gaussian_filter2d
```
Move this imports to the end. Alphabetically.

ghosalsattam replied:
Can you please elaborate?
Earlier I was asked to put alpahbetically(at the present position).
See here-https://github.com/tensorflow/addons/pull/1450#discussion_r412414320

AakashKumarNain replied:
I didn't mean at the bottom. Sorry for incomplete comment.  IMO, it should be like this:
```python
....

from tensorflow_addons.image.dense_image_warp import interpolate_bilinear
from tensorflow_addons.image.interpolate_spline import interpolate_spline
from tensorflow_addons.image.filters import gaussian_filter2d  ## ideally it should be placed here 
from tensorflow_addons.image.filters import mean_filter2d
from tensorflow_addons.image.filters import median_filter2d

....



```

ghosalsattam replied:
@AakashKumarNain  Done.

ghosalsattam commented:
@seanpmorgan Is the PR in a state to bemerged now?

bhack commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+        image = _pad(
+            image, filter_shape, mode=padding, constant_values=constant_values,
+        )
+        conv_ops_x = tf.nn.depthwise_conv2d(
```
Are we sure that apply separable filters it is faster down the stack?
Cause I see Google internals use a single call https://github.com/google/automl/blob/master/efficientdet/aug/autoaugment.py#L1127

ghosalsattam replied:
@bhack  The link in which you have shown uses a very small kernel(3*3). So there it is no effect on the speed(therefore complexity to program is seen).
I have seen other libraries-https://github.com/scipy/scipy/blob/v1.4.1/scipy/ndimage/filters.py#L220-L303
They use multidimentional convolution as sequence of 1d convolution.
The 2d direct convolution requires more multiplication than sequence of 1d convolution.
Consider a (3*3) matrix and a (3*3) kernel. The regular 2d convolution requires 9 multiplication and 9 addition per sample. The sequence of 1d convolutions is like this-6 multiplication(3 for x and y direction each) and 4 additions. Clearly, the difference would increase if kernel_size increases. 
I allowed the user to pass the kernel size which may be large.

bhack replied:
Are you sure that are not separated internally?

ghosalsattam replied:
All 2d convolution_filters are not separable(and the depthwise_conv2d is made keeping in mind the generality).
The gaussian filter is separable(since e^(x^2+y^2)=e^(x^2).e^(y^2)).
I have gone through the code and documentation. It implements as normal(to cope with non separable convolution_filters).

ghosalsattam replied:
Yes separation is not performed internally. As all conv_filters are not separable.
For gaussian filter(e**((x**2)+(y**2))=e(x**2).e**(y**2)). Thus separable convolution filters can also be made.

bhack replied:
Yes I meant that it just seemed strange that we didn't have a flag/arg to tell that the filter are separabile for handling this internally. Also have we any significant overhead with smaller kernels with two API calls?

ghosalsattam replied:
Can you please elaborate the last point?

bhack replied:
I meant if for quite pratical kernel i.e. <= 7x7 it create less overhead a single call with a 2d filter.

ghosalsattam replied:
@bhack The gaussian filter is convolved with an image(which generally have large dimensions). So IMO there would be effect of applying simple gaussian kernel and sequence of 2 1d kernels. Besides we dont know in which situations, the gaussian kernel is used. The user may put large values of the kernel_size. Besides all other libraries-scipy,opencv,skimage uses this technique.
Also computing 2d kernels will cost the python code two for loops(presently all things are handled internally by tensorflow) which will slow down the speed.

WindQAQ replied:
Frankly, no offense, not just to **say** the performance but to **measure** the performance. The overhead of launching tensorflow kernel is not negligible.

ghosalsattam replied:
@WindQAQ can you elaborate? I cannot get you.

ghosalsattam replied:
As far as I can understand you are wanting mathematical performance.
consider the image of size M * N {1024 * 768 (Grayscale)} and kernel =(k * k)(7 * 7)
Computational complexity: Convolution -> O(MNkk)
Computational complexity: Separable convolution -> O(2∗MNk)
Using normal convolution you got O(1024*768*7*7) = 38535168 operations.
Using separable convolution you got(2*1024*768*7) = 11010048 operations.
I m sorry if i'm mistaken to understand you. Please clear if I am mistaken.

bhack replied:
@ghosalsattam I think that generally It could be fast enough to handle the two kernel launch in the most common used image kernel size.
What @WindQAQ is talking about is https://www.tensorflow.org/guide/profiler (See Performance summary point)

WindQAQ replied:
I mean it's better to see some numeric results that compare two 1D separable filters with single 2D filter. I know that separable filters  approach are theoretically faster than single filter,. However, because it involves GPU kernel (the call to `depthwise_conv2d`), things might be different. A single thread taking `n x n` operations together is not necessarily slower than single thread taking `n` operations twice consecutively in parallel computing when it comes to the memory coherence and shared memory etc.

https://www.tensorflow.org/guide/profiler
https://www.hpcs.cs.tsukuba.ac.jp/icpp2019/data/posters/Poster17-abst.pdf

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -28,12 +28,10 @@ def _pad(
     constant_values: TensorLike = 0,
 ) -> tf.Tensor:
     """Explicitly pad a 4-D image.
-
```
Please do not remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -28,12 +28,10 @@ def _pad(
     `tf.nn.depthwise_conv2d`, but supports non-zero, reflect and symmetric
     padding mode. For the even-sized filter, it pads one more value to the
     right or the bottom side.
-
```
Please do not remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -64,7 +62,6 @@ def mean_filter2d(
     name: Optional[str] = None,
 ) -> tf.Tensor:
     """Perform mean filtering on image(s).
-
```
Please do not remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -132,7 +129,6 @@ def median_filter2d(
     name: Optional[str] = None,
 ) -> tf.Tensor:
     """Perform median filtering on image(s).
-
```
Please do not remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+) -> FloatTensorLike:
+    """
+    This function is responsible for having Gaussian Blur. It takes the image as input, computes a gaussian-kernel
+    which follows normal distribution then convolves the image with the kernel. It is implemented as 2 1D convolutions.
```
Please do not insert the empty line below.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+    image: A tensor of shape
+        (batch_size, height, width, channels)
+        (NHWC), (batch_size, channels, height, width)(NCHW).
+
```
Please remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+          G(x,y)=1/(2*3.14*sigma**2)e^((x**2+y**2)/2sigma**2)
+          In 1D,
+          G(x)=e^(-x**2)/2*sigma**2
+
```
Please remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+    filter_shape:An `integer` or `tuple`/`list` of 2 integers, specifying
+        the height and width of the 2-D gaussian filter. Can be a single integer
+        to specify the same value for all spatial dimensions.
+
```
Please remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+        The type of padding algorithm to use, which is compatible with
+        `mode` argument in `tf.pad`. For more details, please refer to
+        https://www.tensorflow.org/api_docs/python/tf/pad.
+
```
Please remove the empty line.

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+    """
+
+    with tf.name_scope(name or "gaussian_filter2d"):
+        if sigma == 0:
```
typically, we also don't want sigma < 0, right? (std should be always positive, though I know it doesn't matter because we take the square on it.)

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+    Raises:
+        Value error if:
+            1). Sigma=0
+            2). passing some string other than ["REFLECT", "CONSTANT", "SYMMETRIC"] in padding.
```
Please see how we format the docstring in the same file. Like `filter_shape: A` (note the space after colon) and the indentation.
```python
"""Main description here.

Args:
  image: AAAAA,
    BBBB.
  filter_shape: CCCC.
"""
```

bhack commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+
+
+@tf.function
+def gaussian_filter2d(
```
Just for note we have already a gaussian kernel creator in Tensorflow but Is private https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py#L3426

ghosalsattam commented:
@bhack @WindQAQ Sorry for delayed response. 
I had run the program on GPU and noted the time for 1D-sequence and 2D convolution. In most of the cases 2D Convolution was faster.
Here are few outputs:
filter size       Image size                     1d-Sequence time              2d filter time(s)
(5,5)            (20,1600,1600,1)                 4.25  s                                  4.12
(5,5)            (20,1600,1600,1)                .00065 s                                .00073
(7,7)            (20,1600,1600,1)                 2.91   s                                  2.49
(7,7)            (20,1600,1600,1)                 .00045 s                               .00043
So I have updated the program. If the program is running on GPU then 2D convolution technique is used. Else the sequence of 1D convolution is used. I am sorry that I earlier did not have the knowledge of the working of GPU. Thanks for your support.

bhack commented:
And if is It TPU?

ghosalsattam commented:
@bhack For that what are you suggesting? Shall I pass a parameter that shall act as a checker to use '2D kernel' or sequence of 1D kernel. I dont have that experience with TPUs and I surfed the net, but got no valuable article to check if tenorflow is running on TPU. Can you please help me out? Thank you

bhack commented:
Check https://colab.research.google.com/notebooks/tpu.ipynb

ghosalsattam commented:
@bhack I have gone through your code. From the code it seems that I have to pass a variable_flag responsible for passing device hint(since the tpu variable would be created somewhere else, if at all created). Shall I pass a new parameter giving device information?

ghosalsattam commented:
@WindQAQ @bhack What are your suggestions on passing an extra parameter to choose the type of convolution?

bhack commented:
I have the impression the we don't need to handle all these things. But in some way to inform `DepthwiseConv2D` that we have a `rank=1` kernel param and all the other stuff needs to be handled by the compilation/transformation stack. 

I don't know why we don't have this interface.

ghosalsattam commented:
So as per now what should be done? Leave it as it is or pass another parameter to select the type of convolution(as done in case of switching between kmeans and kmeans++ algorithm).

bhack commented:
@karmel Do you know anyone internal that could comment about why the `DepWhiseConv2d` don't handle the kernel rank=1 case extra parameter for down the stack compiler/transformations?

@ghosalsattam in the meantine can you try to benchmark It again activating `experimental_compile` param in `tf.function`?

ghosalsattam commented:
@bhack experimental_compile activation is not recommended for public APIs in tensorflow/addons. I am reverting the changes.

bhack commented:
> experimental_compile activation is not recommended for public APIs in tensorflow/addons

@ghosalsattam I don't find a reference to this.

But in any case it was just to see your new results as in https://github.com/tensorflow/addons/pull/1450#issuecomment-620530971

ghosalsattam commented:
@bhack check addons/tools/testing/source_code_test.py::test_no_experimental_api

karmel commented:
> @karmel Do you know anyone internal that could comment about why the DepWhiseConv2d doesn't handle the kernel rank=1 case extra parameter for down the stack compiler/transformations?

@reedwm or @sanjoy - can you advise, or recommend someone else?

bhack commented:
> check addons/tools/testing/source_code_test.py::test_no_experimental_api

@ghosalsattam Check also @alextp post at https://github.com/tensorflow/addons/issues/1752#issuecomment-621322109

ghosalsattam commented:
@bhack  I had checked the GPU time as said by you(in experiment_compile=True). It took .9sec for 1D convolution and .6sec for 2D convolution for shape(1,1600,1600,1).Please provide a review.

alextp commented:
@ghosalsattam and how do those compare with the hand-written kernel times? And are you excluding warm-up costs (measuring warm-up costs for both approaches independently would also be nice)

bhack commented:
@alextp I don't think that we had an hand-written kernel here for gaussian blur. We are just talking about a twice call to `DepWhiseConv2d` TF ops when we have rank=1 kernel (separable filter) vs a single call to `DepWhiseConv2d` with 2d kernel. 

I suppose that is would be hard to fuse on trace two DepWhiseConv2d on the separable filter right?
So I was asking why the `DepWhiseConv2d` in TF doesn't handle the kernel rank=1 case with an parameter for down the stack compiler/transformations?

alextp commented:
Ah, sorry, I misread the thread.

@rmlarsen do you know who understands our depthwise conv kernels enough to answer this?

rmlarsen commented:
Somehow I cannot comment on that thread? Here is what I would like to say:

"I believe @bhack is correct that this could be accomplished with two
1-dimensional convolutions, which would reduce the number of flops.
However, if the kernel is small and the image large, reading the image
twice might slow down the overall computation, compared to using a single
Conv2D call. We should really have a specialized op for separable kernels
like this. (Do we?)"

On Wed, Apr 29, 2020 at 10:25 AM Alexandre Passos <notifications@github.com>
wrote:

> Ah, sorry, I misread the thread.
>
> @rmlarsen <https://github.com/rmlarsen> do you know who understands our
> depthwise conv kernels enough to answer this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/addons/pull/1450#issuecomment-621352450>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA72DUQWBZTWI4GSSMS2BTRPBPJLANCNFSM4LUXGCNA>
> .
>

rmlarsen commented:
Correct me if I'm wrong, but from my experience, I think there are (at
least) 3 distinct regimes for separable kernels (assuming a large image):

1. Tiny kernels: Use brute-force conv2d.
2. Medium kernels: Use "brute force" kernel, but take advantage of
separability, one block at a time.
3. Large kernels: Use FFT.

On Wed, Apr 29, 2020 at 10:34 AM Rasmus Munk Larsen <rmlarsen@google.com>
wrote:

> Somehow I cannot comment on that thread? Here is what I would like to say:
>
> "I believe @bhack is correct that this could be accomplished with two
> 1-dimensional convolutions, which would reduce the number of flops.
> However, if the kernel is small and the image large, reading the image
> twice might slow down the overall computation, compared to using a single
> Conv2D call. We should really have a specialized op for separable kernels
> like this. (Do we?)"
>
> On Wed, Apr 29, 2020 at 10:25 AM Alexandre Passos <
> notifications@github.com> wrote:
>
>> Ah, sorry, I misread the thread.
>>
>> @rmlarsen <https://github.com/rmlarsen> do you know who understands our
>> depthwise conv kernels enough to answer this?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tensorflow/addons/pull/1450#issuecomment-621352450>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEA72DUQWBZTWI4GSSMS2BTRPBPJLANCNFSM4LUXGCNA>
>> .
>>
>

bhack commented:
@rmlarsen I don't know if also the device specific knowledge it is going to play a role in the transform stack.

bhack commented:
@rmlarsen P.s. I don't know if we have already this device specific knowledge or we need to waiting for something like a [more advanced machine retargetability in MLIR](https://llvm.discourse.group/t/rfc-enhancing-machine-retargetability-in-mlir/885)

rmlarsen commented:
Yes, the device will definitely play a role and, if nothing else, determine
where the boundaries of the regimes are.

On Wed, Apr 29, 2020 at 11:31 AM bhack <notifications@github.com> wrote:

> @rmlarsen <https://github.com/rmlarsen> P.s. I don't know if we have
> already this device specific knowledge or we need to waiting for something
> like a more advanced machine retargetability in MLIR
> <https://llvm.discourse.group/t/rfc-enhancing-machine-retargetability-in-mlir/885>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/addons/pull/1450#issuecomment-621386365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA72DTYKIVGATV6UVTUNWTRPBXAVANCNFSM4LUXGCNA>
> .
>

bhack commented:
@rmlarsen @alextp Do you think it makes sense to track part of this in a new issue on the TF repo or is it not interesting enough for other use cases?

ghosalsattam commented:
@bhack For now what should I do to this PR in order for it to be brought to merging state?

bhack commented:
@ghosalsattam I think that you could use the 2d kernel here for merging and that you could track this on a `DepthwiseConv2D` specific issue on the TF repo after waiting for TF team feedback to see if the topic has a "general enough public".

ghosalsattam commented:
And for CPU?

bhack commented:
@ghosalsattam I think you can ignore the device and that it is better that `DepthwiseConv2D` will have directly this knowledge instead of having conditions here in the op so it knows how to propagate down the stack.
In proposal ticket on TF I hope that you could ask to pass the rank=1/separability arg directly (or something similar to inform it about the separate filter).

ghosalsattam commented:
Ok!!!

On Thu, 30 Apr 2020 at 00:42, bhack <notifications@github.com> wrote:

> @ghosalsattam <https://github.com/ghosalsattam> I think you can ignore
> the device and that it is better that DepthwiseConv2D will have directly
> this knowledge instead of having conditions here in the op so it knows how
> to propagate down the stack.
> In proposal ticket on TF I hope that you could ask to pass the
> rank=1/separability arg directly (or something similar to inform it about
> the separate filter).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/addons/pull/1450#issuecomment-621407601>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMBF2CHLHIFYJ5HJS2QNMTRPB3ZXANCNFSM4LUXGCNA>
> .
>

ghosalsattam commented:
@bhack CAn it be merged now?

bhack commented:
You need to wait for a reviewer with write power.

ghosalsattam commented:
@WindQAQ  Can it be merged now?

WindQAQ commented on tensorflow_addons/image/filters.py:
```diff
@@ -204,3 +204,91 @@ def median_filter2d(
+            filter=gaussian_filter_2d,
+            strides=(1, 1, 1, 1),
+            padding="VALID",
+        )
```
Seems that we forget to do `from_4D_image` to retain the original dimension.
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/filters.py#L122

WindQAQ commented on tensorflow_addons/image/tests/filters_test.py:
```diff
@@ -18,6 +18,9 @@
 from tensorflow_addons.image import median_filter2d
+from tensorflow_addons.image import gaussian_filter2d
+from scipy.ndimage import gaussian_filter
+from skimage.filters import gaussian
```
Though it does not matter, I wonder why we use two different implementations to test ours. Can we just use one of them?

ghosalsattam commented:
@WindQAQ I have addressed all your changes. Can it be merged now?

bhack commented:
@rmlarsen @alextp The feature request for Tensorflow is at https://github.com/tensorflow/tensorflow/issues/39050. Please close it if you evaluate that it is not relevant enough.

ghosalsattam commented:
@bhack According to me this should be implemented in all other conv ops where this situation may arise like along wiith depthwise conv operation-  Conv2D, Conv3D,...

ghosalsattam commented:
@all I learnt many new things from this contribution and cant wait for having new contributions. Thanx for your support.

rmlarsen commented:
@ghosalsattam Thanks for the contribution!
